File: examples/sub_module.cpp
#include "example_common.h"
#include "iroha/iroha.h"

using namespace iroha;

IModule *create_module(IDesign *design, const string &name) {
  IModule *mod = new IModule(design, name);
  ITable *table = new ITable(mod);
  mod->tables_.push_back(table);
  IState *st1 = new IState(table);
  table->states_.push_back(st1);
  table->SetInitialState(st1);
  return mod;
}

IModule *create_sub_module(IDesign *design) {
  IModule *mod = create_module(design, "M_sub");
  ITable *tab = mod->tables_[0];
  IResource *res = DesignTool::CreateTaskResource(tab);
  IInsn *insn = new IInsn(res);
  IState *st1 = tab->states_[0];
  st1->insns_.push_back(insn);
  return mod;
}

IModule *create_root_module(IModule *sub_module) {
  IModule *mod = create_module(sub_module->GetDesign(), "M_top");
  ITable *tab = mod->tables_[0];
  IResource *res =
    DesignTool::CreateSubModuleTaskCallResource(tab, sub_module->tables_[0]);
  IInsn *insn = new IInsn(res);
  IState *st1 = tab->states_[0];
  st1->insns_.push_back(insn);
  return mod;
}

IDesign *build_design() {
  IDesign *design = new IDesign;
  IModule *sub = create_sub_module(design);
  IModule *root = create_root_module(sub);
  sub->SetParentModule(root);
  design->modules_.push_back(root);
  design->modules_.push_back(sub);
  DesignTool::Validate(design);
  return design;
}

File: examples/sub_module.py
import sys
sys.path.append('../py')

from iroha import *
from iroha.iroha import *

d = IDesign()
mod_top = IModule(d, "M_top")
tab_top = ITable(mod_top)

mod_sub = IModule(d, "M_sub")
mod_sub.parent_module = mod_top
tab_sub = ITable(mod_sub)

# Callee.
callee_task = design_tool.CreateSubModuleTask(tab_sub)
entry_insn = IInsn(callee_task)
st10 = IState(tab_sub)
st10.insns.append(entry_insn)
tab_sub.states.append(st10)
tab_sub.initialSt = st10

# Caller.
call = design_tool.CreateSubModuleTaskCall(tab_top, tab_sub)
call_insn = IInsn(call)
st20 = IState(tab_top)
st20.insns.append(call_insn)
tab_top.states.append(st20)
tab_top.initialSt = st20


design_tool.ValidateIds(d)
DesignWriter(d).Write()
